lib/puppet/provider/package_targetable.rb

```diff
@@ -0,0 +1,55 @@
+# Targetable package providers implement a 'command' attribute.
+#
+# The packages hash passed to prefetch() is deduplicated, as it is keyed only by name (in transaction.rb).
+# The packages hash should be keyed by all namevars.
+# Instead ...
+#
+# Inspect any package to access the catalog (every package includes a reference to the catalog).
+# Inspect the catalog to find all of the commands for all of the packages of this class.
+# Find all of the package instances using each command including the default of the provider command.
+# Assign each instance's provider by selecting it from the packages hash, based upon name and command,
+
+class Puppet::Provider::Package::Targetable < Puppet::Provider::Package
+  # Prefetch our package list, yo.
+  def self.prefetch(packages)
+    catalog_packages = packages.first[1]::catalog::resources.select{ |p| p.provider.class == self }
+    package_commands = catalog_packages.map { |catalog_package| catalog_package::original_parameters[:command] }.uniq
```

Comment:
As I mentioned in the comment on the parameter - it might be nice to have a comment here describing how `original_parameters` makes this work

Reply:
I'll push a commit for this.